engine/engine.go

```diff
@@ -410,36 +411,38 @@ func (e *Engine) publishAdvForIndex(ctx context.Context, contextID []byte, md me
 		md = metadata.New(metadata.Bitswap{})
 	}
 
+	mdBytes, err := md.MarshalBinary()
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	adv := schema.Advertisement{
+		Provider:  e.h.ID().String(),
+		Addresses: e.retrievalAddrsAsString(),
+		Entries:   cidsLnk,
+		ContextID: contextID,
+		Metadata:  mdBytes,
+		IsRm:      isRm,
+	}
+
 	// Get the previous advertisement that was generated
 	prevAdvID, err := e.getLatestAdCid(ctx)
 	if err != nil {
 		return cid.Undef, fmt.Errorf("could not get latest advertisement: %s", err)
 	}
-	var previousLnk schema.Link_Advertisement
+
 	// Check for cid.Undef for the previous link. If this is the case, then
 	// this means there is a "cid too short" error in IPLD links serialization.
-	if prevAdvID == cid.Undef {
+	if prevAdvID != cid.Undef {
 		log.Info("Latest advertisement CID was undefined - no previous advertisement")
-		previousLnk = nil
-	} else {
-		nb := schema.Type.Link_Advertisement.NewBuilder()
-		err = nb.AssignLink(cidlink.Link{Cid: prevAdvID})
-		if err != nil {
-			return cid.Undef, fmt.Errorf("failed to generate link from latest advertisement cid: %s", err)
-		}
-		previousLnk = nb.Build().(schema.Link_Advertisement)
+		prev := ipld.Link(cidlink.Link{Cid: prevAdvID})
+		adv.PreviousID = &prev
 	}
 
-	mdBytes, err := md.MarshalBinary()
-	if err != nil {
+	// Sign the advertisement.
+	if err := adv.Sign(e.key); err != nil {
```

Comment:
nit: `err` already defined
```suggestion
	if err = adv.Sign(e.key); err != nil {
```

Reply:
I intentionally changed that for tighter scope of `err`